byterun/memory.c

```diff
@@ -686,29 +688,255 @@ CAMLexport CAMLweakdef void caml_modify (value *fp, value val)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return malloc(sz);
+
+  struct pool_block *pb = malloc(sz + SIZEOF_POOL_BLOCK);
+  if (pb == NULL) return NULL;
 #ifdef DEBUG
-  memset (result, Debug_uninit_stat, sz);
+  memset(&(pb->data), Debug_uninit_stat, sz);
+  pb->magic = Debug_pool_magic;
 #endif
+
+  /* Linking the block into the ring */
+  pb->next = pool->next;
+  pb->prev = pool;
+  pool->next->prev = pb;
+  pool->next = pb;
+
+  return &(pb->data);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_alloc(asize_t sz)
+{
+  void *result = caml_stat_alloc_noexc(sz);
+  /* malloc() may return NULL if size is 0 */
+  if ((result == NULL) && (sz != 0))
+    caml_raise_out_of_memory();
   return result;
 }
 
-CAMLexport void caml_stat_free (void * blk)
+CAMLexport void caml_stat_free(caml_stat_block b)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return free(b);
+
+  struct pool_block *pb = get_pool_block(b);
+  if (pb == NULL) return;
+
+  /* Unlinking the block from the ring */
+  pb->prev->next = pb->next;
+  pb->next->prev = pb->prev;
+
+  free(pb);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_resize_noexc(caml_stat_block b, asize_t sz)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return realloc(b, sz);
+
+  struct pool_block *pb = get_pool_block(b);
+  struct pool_block *pb_new = realloc(pb, sz + SIZEOF_POOL_BLOCK);
+  if (pb_new == NULL) return NULL;
+
+  /* Relinking the new block into the ring in place of the old one */
+  pb_new->prev->next = pb_new;
+  pb_new->next->prev = pb_new;
+
+  return &(pb_new->data);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_resize(caml_stat_block b, asize_t sz)
 {
-  free (blk);
+  void *result = caml_stat_resize_noexc(b, sz);
+  if (result == NULL)
+    caml_raise_out_of_memory();
+  return result;
 }
 
 /* [sz] is a number of bytes */
-CAMLexport void * caml_stat_resize (void * blk, asize_t sz)
+CAMLexport caml_stat_block caml_stat_calloc_noexc(asize_t num, asize_t sz)
 {
-  void * result = realloc (blk, sz);
+  sz *= num;
+  caml_stat_block result = caml_stat_alloc_noexc(sz);
+  if (result != NULL)
+    memset(result, 0, sz);
+  return result;
+}
+
+CAMLexport caml_stat_string caml_stat_strdup_noexc(const char *s)
+{
+  size_t slen = strlen(s);
+  caml_stat_block result = caml_stat_alloc_noexc(slen + 1);
+  if (result == NULL)
+    return NULL;
+  memcpy(result, s, slen + 1);
+  return result;
+}
+
+CAMLexport caml_stat_string caml_stat_strdup(const char *s)
+{
+  caml_stat_string result = caml_stat_strdup_noexc(s);
+  if (result == NULL)
+    caml_raise_out_of_memory();
+  return result;
+}
+
+CAMLexport caml_stat_string caml_stat_strconcat(int n, ...)
+{
+  va_list args;
+  char *result, *p;
+  size_t len = 0;
+  int i;
+
+  va_start(args, n);
+  for (i = 0; i < n; i++) {
+    const char *s = va_arg(args, const char*);
+    len += strlen(s);
```

Comment:
This is isn't a regression, as `caml_strconcat` was always like that, but I agree that a check is desirable here. Perhaps, it would be better to leave the "overflow hardening of the runtime" to a separate PR? There are lots of similarly suspicious places in the runtime.